### PR TITLE
Add pre-propose-approval support for multiple choice proposals, and freeze approver

### PR DESCRIPTION
**Pull request by noahsaso**
this does two things:

1. it creates a dao-pre-propose-approval-multiple contract so that multiple choice proposal modules can have approvers, just like single choice proposal modules. it also ensures the approver contract works for both approval-single and approval-multiple

2. it freezes the `approver` in the proposal so that a previously created pending proposal does not change approvers if the approver gets updated while it's still pending. this maintains consistency with other DAO logic to keep things intuitive, where the members who can vote on a proposal are frozen upon proposal creation.

### CODE REVIEW

File: contracts/pre-propose/dao-pre-propose-approval-multiple/src/state.rs
Comment (bekauz) [inline contracts/pre-propose/dao-pre-propose-approval-multiple/src/state.rs:17]: nice!

File: contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs
Comment (bekauz) [inline contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs]: wonder from how many props this could fail. maybe worth to do some test runs?

**Review by bekauz — APPROVED** [review]
this looks great, I don't have much to add.

maybe except that the `dao_pre_propose_base` design combined with message extensions is super clean and holding up really well!

File: contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs
Comment (noahsaso) [reply contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs]: yeah.. what's the easiest way to test this?

Comment (bekauz) [reply contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs]: both test-tube and interchaintest provide this. probably interchaintest would be easiest to test across multiple chains, but require a bit more upfront setup than test-tube

Comment (noahsaso) [reply contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs]: ok but do we really need this? we don't have a rigorous upgrade flow, and the migration will just fail anyways if it runs out of gas. imposing an arbitrary limit serves no real purpose

Comment (noahsaso) [reply contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs]: alternatively, we could make the approver field optional, and if it's empty, default to the currently configured approver! that way, it's backwards compatible and requires no migration 🙌

Comment (noahsaso) [reply contracts/pre-propose/dao-pre-propose-approval-single/src/contract.rs]: > alternatively, we could make the approver field optional, and if it's empty, default to the currently configured approver! that way, it's backwards compatible and requires no migration 🙌

actually nevermind. if any of the types change serialization in the future, not having a migration flow that force-serializes with new types can lead to unparseable state. dang.

**Pull request merged**
